File: scripts/node/parsers/approvalProcessParser.js
'use strict';

const fs = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

/**
 * Reads Approval Process metadata.
 * @param {string} typeDir - path to the approvalProcesses/ folder
 * @returns {{ processes: Array }}
 */
async function parseApprovalProcesses(typeDir) {
  const processes = [];
  if (!fs.existsSync(typeDir)) return { processes };

  const files = fs.readdirSync(typeDir).filter(f => f.endsWith('.approvalProcess-meta.xml'));

  for (const file of files) {
    const xml = fs.readFileSync(path.join(typeDir, file), 'utf8');
    const fullName = file.replace('.approvalProcess-meta.xml', '');
    const entity = fullName.split('.')[0];

    processes.push({
      fullName,
      entity,
      label: extractTag(xml, 'label'),
      active: extractTag(xml, 'active') === 'true',
      description: extractTag(xml, 'description'),
    });
  }

  return { processes };
}

module.exports = { parseApprovalProcesses };

File: scripts/python/parsers/approval_process_parser.py
"""
approval_process_parser.py
Reads Salesforce Approval Process metadata from SFDX source format.
Parses .approvalProcess-meta.xml files.
"""

import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_approval_processes(type_dir: str) -> dict:
    """Parse all *.approvalProcess-meta.xml files in an approvalProcesses/ folder."""
    processes = []
    root = Path(type_dir)
    if not root.exists():
        return {'processes': processes}

    for file in sorted(root.glob('*.approvalProcess-meta.xml')):
        xml = file.read_text(encoding='utf-8')
        full_name = file.name.replace('.approvalProcess-meta.xml', '')
        entity = full_name.split('.')[0]

        processes.append({
            'fullName': full_name,
            'entity': entity,
            'label': _extract_tag(xml, 'label'),
            'active': _extract_tag(xml, 'active') == 'true',
            'description': _extract_tag(xml, 'description'),
        })

    return {'processes': processes}
